dockerfile-image-update/src/test/java/com/salesforce/dockerfileimageupdate/utils/RateLimiterTest.java

```diff
@@ -0,0 +1,172 @@
+package com.salesforce.dockerfileimageupdate.utils;
+
+import com.google.common.collect.ImmutableMap;
+import io.github.bucket4j.TimeMeter;
+import java.time.Duration;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.Callable;
+import java.util.concurrent.FutureTask;
+import java.util.concurrent.atomic.AtomicInteger;
+import net.sourceforge.argparse4j.inf.Namespace;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+import static org.testng.Assert.*;
+
+public class RateLimiterTest {
+
+    @DataProvider(name = "rateLimitingData")
+    public static Object[][] rateLimitingData() {
+        return new Object[][]{{2, 10, 5}, {4, 16, 4}};
+    }
+
+    @DataProvider(name = "getRateLimitingData")
+    public static Object[][] getRateLimitingData() {
+        return new Object[][]{
+                {"1234-per-2h", RateLimiter.class, false},
+                {"500", RateLimiter.class, false},
+                {"500-per-s", RateLimiter.class, false},
+                {"500-per-60s", RateLimiter.class, false},
+                {"image", null, true},
+                {"500-random", null, true},
+                {"", null, true},
+                {"null", null, true}
+        };
+    }
+
+    /**
+     * This test is to test the rate Limiter against the wall clock.
+     * This will run at least the rate limit seconds.
+     */
+    @Test(dataProvider = "rateLimitingData")
+    public void testingAgainstWallClock(int rateLimit, int durationLimit, int tokenAddingRate) throws Exception {
+        FutureTask futureTask = new FutureTask(new RateLimiterCallable
+                (new RateLimitWrapper(null, rateLimit, durationLimit, tokenAddingRate)));
+        Thread t1 = new Thread(futureTask);
+        t1.start();
+        Thread.sleep(100);
+        // the token should be consumed within first 3s leaving the thread in blocked state as it'd be waiting for
+        // next token to be available
+        assertEquals(t1.getState().toString(), Thread.State.TIMED_WAITING.toString());
+        t1.join();
+        assertEquals(t1.getState().toString(), Thread.State.TERMINATED.toString());
+        assertEquals(futureTask.get(), Optional.ofNullable(null));
+    }
```

Comment:
Does this cover any use case that `testingAgainstCustomClock` doesn't?

Reply:
Testing against a real system clock. that depends on the current time in Millis instead of the user-provided clock. Make sense to still have it while all other tests depends on the custom clock